File: go/memorysync/rewriter.go
package memorysync

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// StripTypes are event types with no uuid, not in the parentUuid DAG
// (Step 0 verified: stripping them cannot break the resume chain).
var StripTypes = map[string]bool{
	"file-history-snapshot": true,
	"queue-operation":       true,
}

// Sub is one origin->target path substitution.
type Sub [2]string

// BuildSubs returns ordered origin->target path substitutions (most-specific
// first; bare dashed name last as a catch-all).
func BuildSubs(srcCwd, dstCwd string) []Sub {
	home, _ := os.UserHomeDir()
	srcD := Dashed(srcCwd)
	dstD := Dashed(dstCwd)
	srcReal, err := filepath.EvalSymlinks(srcCwd)
	if err != nil {
		srcReal = srcCwd
	}
	dstReal, err := filepath.EvalSymlinks(dstCwd)
	if err != nil {
		dstReal = dstCwd
	}
	return []Sub{
		{filepath.Join(home, ".claude", "projects", srcD), filepath.Join(home, ".claude", "projects", dstD)},
		{srcReal, dstReal},
		{srcCwd, dstCwd},
		{srcD, dstD},
	}
}

// RewriteStr applies the substitutions to a string.
func RewriteStr(s string, subs []Sub) string {
	for _, sub := range subs {
		if sub[0] != "" && strings.Contains(s, sub[0]) {
			s = strings.ReplaceAll(s, sub[0], sub[1])
		}
	}
	return s
}

// RewriteObj recursively rewrites all strings in o (str/list/map).
func RewriteObj(o any, subs []Sub) any {
	switch v := o.(type) {
	case string:
		return RewriteStr(v, subs)
	case []any:
		for i, x := range v {
			v[i] = RewriteObj(x, subs)
		}
		return v
	case map[string]any:
		for k, val := range v {
			v[k] = RewriteObj(val, subs)
		}
		return v
	}
	return o
}

// RewriteStats holds the Step-0 invariants from a rewrite pass.
type RewriteStats struct {
	Kept             int
	Stripped         int
	Orphans          int
	ResidueSrcCwd    int
	ResidueSrcDashed int
	Malformed        int
}

// RewriteSession reads src JSONL, strips machine-local events, rewrites paths,
// writes dst JSONL. Returns stats incl. orphan + residue checks.
func RewriteSession(srcJSONL, dstJSONL, srcCwd, dstCwd string) (RewriteStats, error) {
	subs := BuildSubs(srcCwd, dstCwd)
	srcD := Dashed(srcCwd)
	dstD := Dashed(dstCwd)
	var stats RewriteStats
	in, err := os.Open(srcJSONL)
	if err != nil {
		return stats, err
	}
	defer in.Close()
	out, err := os.Create(dstJSONL)
	if err != nil {
		return stats, err
	}
	scanner := bufio.NewScanner(in)
	scanner.Buffer(make([]byte, 1024*1024), 16*1024*1024) // for long lines
	enc := json.NewEncoder(out)
	enc.SetEscapeHTML(false)
	var objs []map[string]any
	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}
		var o map[string]any
		if err := json.Unmarshal([]byte(line), &o); err != nil {
			stats.Malformed++
			continue
		}
		t, _ := o["type"].(string)
		if StripTypes[t] {
			stats.Stripped++
			continue
		}
		o = RewriteObj(o, subs).(map[string]any)
		enc.Encode(o)
		objs = append(objs, o)
		stats.Kept++
	}
	out.Close()
	if err := scanner.Err(); err != nil {
		return stats, err
	}
	// orphan check
	uuids := map[string]bool{}
	for _, o := range objs {
		if u, ok := o["uuid"].(string); ok {
			uuids[u] = true
		}
	}
	for _, o := range objs {
		if p, ok := o["parentUuid"].(string); ok && p != "" && !uuids[p] {
			stats.Orphans++
		}
	}
	// residue check (dst-span masking for both cwd + dashed forms)
	data, _ := os.ReadFile(dstJSONL)
	for _, line := range strings.Split(string(data), "\n") {
		if strings.Contains(line, srcCwd) {
			masked := line
			if dstCwd != "" && strings.Contains(dstCwd, srcCwd) {
				masked = strings.ReplaceAll(line, dstCwd, strings.Repeat("\x00", len(dstCwd)))
			}
			if strings.Contains(masked, srcCwd) {
				stats.ResidueSrcCwd++
			}
		}
		if strings.Contains(line, srcD) {
			masked := line
			if dstD != "" && strings.Contains(dstD, srcD) {
				masked = strings.ReplaceAll(line, dstD, strings.Repeat("\x00", len(dstD)))
			}
			if strings.Contains(masked, srcD) {
				stats.ResidueSrcDashed++
			}
		}
	}
	return stats, nil
}

File: python/memory_sync/rewriter.py
"""Path rewriting + event taxonomy for Claude Code session JSONL."""
import json
import os
from dataclasses import dataclass

from . import encoding

# Event types with no `uuid` and not in the parentUuid DAG (Step 0 verified:
# 0 orphans; stripping them cannot break the resume chain).
STRIP_TYPES = {"file-history-snapshot", "queue-operation"}


def build_subs(src_cwd: str, dst_cwd: str) -> list[tuple[str, str]]:
    """Ordered origin->target path substitutions.

    Most-specific first; the bare dashed project-dir name is last as a catch-all.
    M2 uses direct same-HOME substitution (HOME is the SAME on both sides);
    cross-HOME logical-placeholder resolution is deferred to the Go port.
    """
    home = os.path.expanduser("~")
    src_d = encoding.dashed(src_cwd)
    dst_d = encoding.dashed(dst_cwd)
    return [
        (os.path.join(home, ".claude", "projects", src_d),
         os.path.join(home, ".claude", "projects", dst_d)),
        (os.path.realpath(src_cwd), os.path.realpath(dst_cwd)),
        (src_cwd, dst_cwd),
        (src_d, dst_d),
    ]


def rewrite_str(s, subs) -> str:
    if not isinstance(s, str):
        return s
    for a, b in subs:
        if a and a in s:
            s = s.replace(a, b)
    return s


def rewrite_obj(o, subs):
    if isinstance(o, str):
        return rewrite_str(o, subs)
    if isinstance(o, list):
        return [rewrite_obj(x, subs) for x in o]
    if isinstance(o, dict):
        return {k: rewrite_obj(v, subs) for k, v in o.items()}
    return o


@dataclass
class RewriteStats:
    kept: int = 0
    stripped: int = 0
    orphans: int = 0
    residue_src_cwd: int = 0
    residue_src_dashed: int = 0
    malformed: int = 0


def rewrite_session(src_jsonl: str, dst_jsonl: str, src_cwd: str, dst_cwd: str) -> RewriteStats:
    """Read src JSONL, strip machine-local events, rewrite paths, write dst JSONL.

    Returns stats incl. orphan-parentUuid and residue checks (the Step 0 invariants).
    Does NOT copy memory/tool-results sideload — that is checkpoint.py's job in M2.
    """
    subs = build_subs(src_cwd, dst_cwd)
    src_d = encoding.dashed(src_cwd)
    dst_d = encoding.dashed(dst_cwd)
    stats = RewriteStats()
    out_objs = []
    with open(src_jsonl) as f, open(dst_jsonl, "w") as out:
        for line in f:
            line = line.rstrip("\n")
            if not line.strip():
                continue
            try:
                o = json.loads(line)
            except json.JSONDecodeError:
                stats.malformed += 1
                continue
            t = o.get("type", "?")
            if t in STRIP_TYPES:
                stats.stripped += 1
                continue
            o = rewrite_obj(o, subs)
            out.write(json.dumps(o, ensure_ascii=False) + "\n")
            out_objs.append(o)
            stats.kept += 1

    uuids = {o.get("uuid") for o in out_objs if isinstance(o, dict) and o.get("uuid")}
    for o in out_objs:
        if isinstance(o, dict):
            p = o.get("parentUuid")
            if p is not None and p not in uuids:
                stats.orphans += 1

    with open(dst_jsonl) as f:
        for line in f:
            # residue checks: count src NOT covered by a dst span (avoids
            # false-positive when dst contains src as a substring)
            if src_cwd in line:
                masked = line.replace(dst_cwd, "\x00" * len(dst_cwd)) if dst_cwd and src_cwd in dst_cwd else line
                if src_cwd in masked:
                    stats.residue_src_cwd += 1
            if src_d in line:
                masked = line.replace(dst_d, "\x00" * len(dst_d)) if dst_d and src_d in dst_d else line
                if src_d in masked:
                    stats.residue_src_dashed += 1

    return stats
